common/flatpak-run.c

```diff
@@ -1248,6 +1248,49 @@ flatpak_run_add_dconf_args (FlatpakBwrap *bwrap,
   return TRUE;
 }
 
+static gboolean
+flatpak_run_save_environ (GStrv         run_environ,
+                          char          *dir,
+                          GCancellable  *cancellable,
+                          GError       **error)
+{
+  g_autofree char *string = NULL;
+  gsize size = 0;
+  glnx_autofd int dir_fd = -1;
+
+  if (*run_environ)
+    {
+      gsize i;
+      char *ptr;
+
+      /* First part, getting size */
+      for (i = 0; run_environ[i] != NULL; i++)
+        size += strlen (run_environ[i]) + 1;
+
+      /* Second part, building string */
+      string = g_new (char, size);
+      for (i = 0, ptr = string; run_environ[i] != NULL; i++)
+        ptr = g_stpcpy (ptr, run_environ[i]) + 1;
+    }
+
+  dir_fd = openat (AT_FDCWD, dir,
+                   O_PATH | O_DIRECTORY | O_CLOEXEC);
+  if (dir_fd < 0)
+    return glnx_throw_errno_prefix (error,
+                                    _("Unable to open directory %s"),
+                                    dir);
```

Comment:
Use `glnx_opendirat`, or if you really want to compose the error message yourself, `glnx_opendirat_with_errno`. This gets some low-level subtleties right (in particular, use of `O_NOCTTY`, which you haven't done here).

Reply:
This is based on the implementation of the per-app dir, with the assumption that it handled such things correctly:
https://github.com/flatpak/flatpak/blob/106d2065c8d3d819b2713ee5a534231a0c73ff04/common/flatpak-instance.c#L575-L583

Do we have an issue there as well? (`O_NOCTTY` or otherwise)

Reply:
I think yes we do, I should have used `glnx_opendirat` or `glnx_opendirat_with_errno`. Would you mind opening a PR for that too, while you're in this part of the codebase? (Or I could do it, but then I'd have to wait for a reviewer, which could take a long time!)

Reply:
This PR now uses `glnx_opendirat`, but I haven't had a chance to open a PR for the existing issue yet. I will try to do so next week.

Reply:
> I think yes we do, I should have used `glnx_opendirat` or `glnx_opendirat_with_errno`. Would you mind opening a PR for that too, while you're in this part of the codebase? (Or I could do it, but then I'd have to wait for a reviewer, which could take a long time!)

I opened a PR for this: https://github.com/flatpak/flatpak/pull/5527